### ISSUE TITLE
"Storing non-optimized index is not supported yet" for the bit_hamming space in HNSW

### ISSUE THREAD

**Issue by funtusov** [searchivarius/nmslib#119]
Hi! I ran into an unexpected issue when trying out the `bit_hamming` space (and `INT` distance type): a `Storing non-optimized index is not supported yet` exception when trying to save the index. 

During creation `skip_optimized_index` is set to 0

The same code works for `cosinesimil` space and `FLOAT` distance type

Is there an underlying reason why this isn't expected to work for the hamming space, or is it just not implemented yet/a bug?

A potentially relevant note: when creating index, I get the following message:
`hnsw.cc:192 (CreateIndex) [INFO] No appropriate custom distance function for Hamming (bit-storage) space`
At the same time, NN search seems to work reasonably. What exactly does that message mean?

Thanks!

**Comment by funtusov** [comment]
After further reading, the answer, probably, is that there are no optimized index implementations for spaces other than L2 and cosine similarity? Is index saving/loading not supported in those cases then?

**Comment by searchivarius** [comment]
Hi Greg,

yes, this is not fixed yet, unfortunately. When do you need to have it working?

**Comment by funtusov** [comment]
Hi Leonid, not sure yet, we're still investigating potential ways to use the thing :) Is it something that needs to be done space-by-space, or is there a possible general solution for all custom spaces?

**Comment by searchivarius** [comment]
There is surely a general solution. It was postponed for a couple of uninteresting reasons.

**Comment by svirg** [comment]
Hi, I'm also in need for bit_hamming space support for hnsw
Is there a lot of work to be done to implement this?
I consider doing it by myself, but I have no idea of how much effort it will take. Right now it seems for me that there are 2 things to do in order to make it work:
1) add specialized function that calculates hamming distances
2) use contiguous memory layout for storing data points and move unoptimized data to it

However, couple of things bother me
Do I really have to implement hamming distance function? what is the purpose of functions defined in similarity_search/src/method/hnsw_distfunc_opt.cc ? Haven't they been implemented somewhere else?
Will it be tricky to rearrange data in an optimized memory layout?

**Comment by yurymalkov** [comment]
Hi @svirg!
If you do not need index serialization, you can use any distance from the nmslib.
Currently hnsw uses its own element storage for making more fast optimized index; and the hnsw_distfunc_opt.cc file contains just a bit optimized **copies** of L2 and cosine similarity distances of the ones already implemented in the nmslib. 

The major difference is in the input. In nmslib the input is links to "Object" structure which than have links to the actual data, while in hnsw_distfunc_opt.cc the input is the links to the data and object sizes (i.e. no need for intermediate objects and thus no performance loss). The hnsw copies the data to the optimized index only after building the index, and, currently hnsw can save only the optimized version of the index.

There are at least two possible solutions to make hsnw serializable:
1) Write an additional serializer for non-optimized hnsw index, like the one in sw-graph algorithm. This will work for all distances defined in the nmslib, however it will also lead to notable loss of performance for fast distances (though @searchivarius has plans on reviewing the Object class to address this problem).
2) If you want to add a new distance to the optimized version (from the ones that are present in the nmslib), you currently have to copy it from nmslib with slight adjustments to the hnsw_distfunc_opt.cc and add a few references at index building, loading and saving.

**Comment by svirg** [comment]
Yury, thanks for your reply, things are getting clearer
I'm mostly concerned about memory footprint, not the search speed. Does optimized index consumes less RAM that unoptimized one?

**Comment by yurymalkov** [comment]
The optimized index generally uses less memory (possibly because it holds straight memory constraints as in the paper), but it can be usefully only at runtime, since currently you cannot build the optimized index without building a non-optimized version. Moreover, the conversion of indexes is now done in the main memory, requiring even more RAM (this however can be fixed quite easily).

Potentially, one can build only the optimized index (which has straight memory constraints and also should require less time to make), but this would need a lot of rewriting in the construction algorithm.

**Comment by searchivarius** [comment]
@svirg no, I don't think that our current algorithms are will be memory-efficient for bit strings. Out of curiousity,  what kind of data set do you have (#of recs, #of bits)?

**Comment by svirg** [comment]
@searchivarius, our dataset has roughly 1e10 elements
Currently we use sharded solution with cosine distance and 64-d float vectors. We would like to switch to hamming distance and  512-bit vectors. Supposedly, it will allow us to use 3x less space for out indices. We use default value for M (=16)

**Comment by searchivarius** [comment]
@svirg thanks!

**Comment by lsorber** [comment]
This is a feature we would also be very happy to see from nmslib. Specifically, support for saving/loading sparse cosine distance spaces with hnsw. Without sparse support, we can only save (huge) dense models.

**Comment by searchivarius** [comment]
@lsorber I have started working on this, hopefully it won't take longer than a couple of weeks.

**Comment by searchivarius** [comment]
Hi @lsorber could you test saving of the index in the dev version? There seems to be some issue on Windows, but the Linux version should work fine. Thanks!

**Comment by lsorber** [comment]
Hi @searchivarius, I would be glad to help. I have a Windows machine, do you want me to go ahead and test (since you mentioned there are still issues on Windows)?

**Comment by searchivarius** [comment]
@lsorber no, the index saving wasn't supported, but let me test this feature on Windows first.

**Comment by searchivarius** [comment]
This should now work for DEV.
